README
NAME
    once - execute code only once throughout the program's lifetime

VERSION
    version 1.100790

SYNOPSIS
        use once;

        sub setup {
            my $self = shift;
            ONCE {
                print "run only once however often setup() is called"
            };
            # other_things();
        }

DESCRIPTION
    This module provides a way to run code only once, no matter how often
    the surrounding code is called.

METHODS
  "ONCE"
    This subroutine is exported automatically. It takes a code block and
    executes it only the first time that this specific call of "ONCE" is
    encountered. This might be useful, for example, in a situation where you
    initialize an object but only want to do it the first time any object of
    that class is initialized, perhaps to install methods or setup some
    other data.

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=once>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/once/>.

    The development version lives at <http://github.com/hanekomu/once/>.
    Instead of sending patches, please fork this project using the standard
    git and github infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension once

1.100790  2010-03-20 00:09:58 Europe/Vienna
    - converted distribution to Dist::Zilla-style

0.02  Fri Mar 20 09:03:21 CET 2009 (Marcel Gruenauer <marcel@cpan.org>)
     - pod fixes (thanks Miyagawa Tatsuhiko)

0.01  Fri Mar 19 21:38:13 CET 2009 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


lib/once.pm
use 5.008;
use strict;
use warnings;

package once;
our $VERSION = '1.100790';

# ABSTRACT: execute code only once throughout the program's lifetime
use Exporter qw(import);
our @EXPORT  = qw(ONCE);

sub ONCE (&) {
    my $code = shift;
    our %seen;
    my ($package, $filename, $line) = caller;
    unless ($seen{"ONCE $package $filename $line"}++) {
        $code->();
    }
}
1;


__END__
=pod

=head1 NAME

once - execute code only once throughout the program's lifetime

=head1 VERSION

version 1.100790

=head1 SYNOPSIS

    use once;

    sub setup {
        my $self = shift;
        ONCE {
            print "run only once however often setup() is called"
        };
        # other_things();
    }

=head1 DESCRIPTION

This module provides a way to run code only once, no matter how often the
surrounding code is called.

=head1 METHODS

=head2 C<ONCE>

This subroutine is exported automatically. It takes a code block and executes
it only the first time that this specific call of C<ONCE> is encountered. This
might be useful, for example, in a situation where you initialize an object but
only want to do it the first time any object of that class is initialized,
perhaps to install methods or setup some other data.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=once>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/once/>.

The development version lives at
L<http://github.com/hanekomu/once/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'once',
                       'DISTNAME' => 'once',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'execute code only once throughout the program\'s lifetime',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100790',
                       'PREREQ_PM' => {
                                        'Test::More' => '0.88',
                                        'Scalar::Util' => '0',
                                        'Test::Differences' => '0',
                                        'English' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'Exporter' => '0',
                                        'Carp' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



